packages/frontend/src/components/Consensus/ConsensusBlock.tsx

```diff
@@ -205,25 +208,33 @@ export class ConsensusBlock extends React.Component<ConsensusBlock.Props, {}> {
       let statPrecommit;
 
       if (implicitPrevote) {
-        statPrevote = <Icon src={checkIcon} className="implicit" alt="Implicit Prevote"/>;
+        statPrevote = this.props.icons.implicitPrevote;
       }
       if (implicitPrecommit) {
-        statPrecommit = <Icon src={checkIcon} className="implicit" alt="Implicit Precommit"/>;
+        statPrecommit = this.props.icons.implicitPrecommit;
       }
 
       if (prevote) {
-        statPrevote = <Icon src={checkIcon} className="explicit" alt="Prevote"/>;
+        statPrevote = this.props.icons.explicitPrevote;
       }
       if (precommit) {
-        statPrecommit = <Icon src={checkIcon} className="explicit" alt="Precommit"/>;
+        statPrecommit = this.props.icons.explicitPrecommit;
       }
 
       const stat = [statPrevote, statPrecommit];
-      return <Tooltip text={tooltipText}>{stat}</Tooltip>
+      return <span>{stat}</span>;
+    } else {
+      return this.props.icons.self;
+    }
+  }
+
+  private getJdenticon(hash: Types.BlockHash) {
+    if (this.state.jdenticons[this.props.compact ? 0 : 1].hasOwnProperty(hash)) {
```

Comment:
Without the "Reuse icons/identicons/jdenticons" commits I see that for each new block which is appended in the view we have long, recurring calls to these components. With these commits there are no recurring calls and the overall update time is shorter.

In the case of the identicon the number of icons are (basically) fixed, there are as many identicons as validators. My understanding is that without these reuse-commits these identicon components are recreated each time a new block is appended in the view. This is because they are child components of the `ConsensusBlock` component and a new instance of this component is created for each new block. Recreating identicons is especially expensive, since they are displayed in each column and row of the matrix.

The screenshot below is without the reuse-commits.

Wdyt?

![0](https://user-images.githubusercontent.com/241530/58343259-d788de00-7e52-11e9-8a6b-035bd71bd8e7.png)

Reply:
I reverted it for now, we can look into it if we want to optimize further.

Reply:
Rendering bottlenecks should be handled by `shouldComponentUpdate`, I think you can sprinkle it to parent components where you are sure an update needs or needs not happen since you have access to props. As long as the tree is constructed in place, and all the lists are properly done with `key`s that are fixed for components' lifetime, React should never have to re-initialize those.